### PR TITLE
SLVS-1592 Update TaintStore with taints from SLCore

### PR DESCRIPTION
**Pull request by georgii-borovinskikh-sonarsource**
Part of

### CODE REVIEW

**Review by gabriela-trutan-sonarsource — CHANGES_REQUESTED** [review]
Functionality works, but:

- You missed a todo that was supposed to be done in this task in [TaintVulnerabilityListener](https://github.com/SonarSource/sonarlint-visualstudio/blob/e64f8fdea71591678fefa90e6614756274e528c4/src/SLCore.Listeners/Implementation/TaintVulnerabilityListener.cs#L41)
- I have some more feedback

File: src/IssueViz.Security/Taint/TaintStore.cs [hunk 1]
```diff
@@ -61,31 +55,29 @@ internal sealed class TaintStore : ITaintStore
...
         public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
         {
-            lock (Locker)
+            lock (locker)
             {
                 return taintVulnerabilities.ToList();
             }
         }
 
-        public AnalysisInformation GetAnalysisInformation() => analysisInformation;
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintStore.cs:77]: I will fix it in the SSE PR, as the method set is going to change there

File: src/IssueViz.Security/Taint/TaintStore.cs [hunk 2]
```diff
@@ -108,8 +101,15 @@ public void Remove(string issueKey)
...
+            IAnalysisIssueVisualization valueToRemove;
+
+            lock (locker)
             {
+                if (configurationScope == null)
+                {
+                    return;
+                }
+
                 var indexToRemove =
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintStore.cs:113]: 1) This method is removed in SSE PR 2) it is better technically. Remove by index is faster than by value

File: src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs [hunk 3]
```diff
@@ -18,185 +18,173 @@
...
+        return result;
+    }
 
-            var branches = await sonarQubeService.GetProjectBranchesAsync(projectKey, CancellationToken.None);
+    private bool IsConfigScopeReady(ConfigurationScope configurationScope)
+    {
+        var isReady = configurationScope.RootPath is not null;
+        if (!isReady)
+        {
+            logger.LogVerbose("Configuration scope root hasn't been initialized.");
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs:130]: IMO it's not required for verbose logs

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs [hunk 4]
```diff
@@ -301,7 +301,7 @@ private void UpdateIssues()
                 taintIssueViewModel.TaintIssueViz.PropertyChanged += OnTaintIssuePropertyChanged;
             }
 
-            AnalysisInformation = store.GetAnalysisInformation();
+            AnalysisInformation = new AnalysisInformation("stub", DateTimeOffset.Now);
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:304]: https://github.com/SonarSource/sonarlint-visualstudio/pull/5829

File: src/IssueViz.Security/Taint/TaintIssuesConfigurationScopeMonitor.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        private readonly ITaintIssuesSynchronizer taintIssuesSynchronizer;
+
+        [ImportingConstructor]
+        public TaintIssuesConfigurationScopeMonitor(IActiveConfigScopeTracker activeConfigScopeTracker,
+            ITaintIssuesSynchronizer taintIssuesSynchronizer)
+        {
+            this.activeConfigScopeTracker = activeConfigScopeTracker;
+            this.taintIssuesSynchronizer = taintIssuesSynchronizer;
+
+            this.activeConfigScopeTracker.CurrentConfigurationScopeChanged += ActiveConfigScopeTrackerOnCurrentConfigurationScopeChanged;
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintIssuesConfigurationScopeMonitor.cs:47]: Separable responsibility, can be (and already was) in a different class

File: src/IssueViz.Security/Taint/TaintStore.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabriela-trutan-sonarsource) [reply src/IssueViz.Security/Taint/TaintStore.cs:77]: Can you maybe write it in the corresponding task, so we won't forget?

File: src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabriela-trutan-sonarsource) [reply src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs:130]: For consistency reasons, I think we should do it.

File: src/IssueViz.Security/Taint/TaintIssuesConfigurationScopeMonitor.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (gabriela-trutan-sonarsource) [reply src/IssueViz.Security/Taint/TaintIssuesConfigurationScopeMonitor.cs:47]: I know it already has. But does it make sense to have just a class with an event subscription, that delegates the work to the synchronizer?
The responsibility of the synchronizer is to synchronize! So it should know when to do that as well (like subscribing to the corresponding events)

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabriela-trutan-sonarsource) [reply src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:304]: I see. Maybe next time add a todo with a link to the task/PR? :)

File: src/IssueViz.Security/Taint/TaintStore.cs [hunk 6]
```diff
@@ -34,12 +31,9 @@ internal interface ITaintStore : IIssuesStore
...
         /// Can be called multiple times.
         /// </summary>
-        void Set(IEnumerable<IAnalysisIssueVisualization> issueVisualizations, AnalysisInformation analysisInformation);
+        void Set(IReadOnlyCollection<IAnalysisIssueVisualization> issueVisualizations, string newConfigurationScope);
 
-        /// <summary>
-        /// Returns additional analysis information for the existing visualizations in the store.
-        /// </summary>
-        AnalysisInformation GetAnalysisInformation();
+        string ConfigurationScope { get; }
```
[the remark this answers is not in the document]
Comment (georgii-borovinskikh-sonarsource) [reply src/IssueViz.Security/Taint/TaintStore.cs:36]: Discussed. Keeping this to be used in SSE implementation PR to discard events for sse events that happen out of order

**Review by gabriela-trutan-sonarsource — APPROVED** [review]

**Pull request merged**